torch/_inductor/comms.py

```diff
@@ -342,6 +345,198 @@ def reorder_compute_and_comm_for_overlap(
+                if x.alias_info is not None and x.alias_info.is_write
+            ]
+            if isinstance(node.target, torch._ops.OpOverload)
+            else []
+        )
+        mutated_node_arg_storages = {
+            StorageWeakRef(node.args[i].meta["val"].untyped_storage())
+            for i in mutated_arg_idxes
+        }
+        storages_of_unsharded_params = {
+            StorageWeakRef(unsharded_param.meta["val"].untyped_storage())
+            for unsharded_param in unsharded_params
+        }
+        return len(mutated_node_arg_storages & storages_of_unsharded_params) > 0
+
+    for node in node_list:
+        if (
+            node.op == "call_function"
+            and isinstance(node.target, torch._ops.OpOverload)
+            and node.target._schema.is_mutable
+            and not is_allowed_mutation(node)
+        ):
+            assert not is_node_mutating_unsharded_param_or_its_alias(
+                node, unsharded_param_to_fsdp_copy_node_idxes.keys()
+            ), f"""\
+User mutation on FSDP2 unsharded param is not allowed when Traceable FSDP2 is used. Violating node: {node}
+"""
+
+    # For each `fsdp.copy_(unsharded_param, Y)`, replace downstream usage of `unsharded_param` with `Y`.
+    #
+    # NOTE: Because of "layer reuse" use case, there could be multiple `fsdp.copy_` to the same `unsharded_param` graph input.
+    # e.g.
+    # ```
+    #     fsdp_copy_1 = fsdp.copy_(unsharded_param_1, Y1)
+    #     ... (use of unsharded_param_1)                     -> Subgraph 1
+    #     fsdp_copy_2 = fsdp.copy_(unsharded_param_1, Y2)
+    #     ... (use of unsharded_param_1)                     -> Subgraph 2
+    #     fsdp_copy_3 = fsdp.copy_(unsharded_param_1, Y3)
+    #     ... (use of unsharded_param_1)                     -> Subgraph 3
+    # ```
+    # We must do the replacement only within each subgraph.
+    replacement_nodes = set()
+    for (
+        unsharded_param,
+        fsdp_copy_node_idxes,
+    ) in unsharded_param_to_fsdp_copy_node_idxes.items():
+        for i, fsdp_copy_node_idx in enumerate(fsdp_copy_node_idxes):
+            fsdp_copy_node = node_list[fsdp_copy_node_idx]
+            assert fsdp_copy_node.args[0] is unsharded_param
+            _, replacement = fsdp_copy_node.args
+            replacement_nodes.add(replacement)
+            # subgraph_start_idx is exclusive
+            subgraph_start_idx = fsdp_copy_node_idx + 1
+            # subgraph_end_idx is exclusive (also intentionally don't replace args in return op)
+            subgraph_end_idx = (
+                fsdp_copy_node_idxes[i + 1]
+                if i < len(fsdp_copy_node_idxes) - 1
+                else len(node_list) - 1
+            )
+            subgraph_nodes = node_list[subgraph_start_idx:subgraph_end_idx]
+            assert unsharded_param is not None
+            assert not any(
+                is_node_mutating_unsharded_param_or_its_alias(node, [unsharded_param])
```

Comment:
hmm so maybe i'm being too pedantic, but:

(1) this helper fn will technically ignore `fsdp.copy_` and `inductor.storage_resize_bytes_` when figuring out if there are any mutations in this subgraph

(2) that seems bad though, because this swap will be wrong if there are **any** mutations in this subgraph (including an extra `fsdp.copy_` in the middle)

(3) this subgraph represents a region of nodes between two existing `fsdp.copy_` calls, so I guess we shouldn't really expect there to be any extra `fsdp.copy_` nodes inside the region? It's not obvious to me that this is true though, since `fsdp_copy_node_idxes` doesn't contain the index of **every** `fsdp.copy_` nodes - only the ones that we decided are safe to remove.

Reply:
> I guess we shouldn't really expect there to be any extra fsdp.copy_ nodes inside the region?

There can be extra fsdp.copy_ nodes inside the region for the layer reuse case: https://github.com/pytorch/pytorch/pull/133730#discussion_r1755284763.

I believe as long as the mutations are not operating on the same unsharded param (or its alias), they should be benign? We don't have to ban all mutations (and also can't because of the layer reuse case)